Add support for SSL Keystores

tzolov opened the pull request:
- Introduce an GemfireSslProperties (gemfire.security.ssl.*)
  - Extends GemfireClientCacheConfiguration to covert the GemfireSslProperties into native Gemfire/Geode SSL properties.
  - Added SslGemfireSinkIntegrationTests test for the sink gemfire app. But had to disable the existing GemfireSinkIntegrationTests due to test infrastructure racing conditions.

  Resolve #12

tzolov committed:
Add support for SSL Keystores

  - Introduce an GemfireSslProperties (gemfire.security.ssl.*)
  - Extends GemfireClientCacheConfiguration to covert the GemfireSslProperties into native Gemfire/Geode SSL properties.
  - Added SslGemfireSinkIntegrationTests test for the sink gemfire app. But had to disable the existing GemfireSinkIntegrationTests due to test infrastructure racing conditions.

  Resolve #12

artembilan commented on spring-cloud-starter-stream-common-gemfire/src/main/java/org/springframework/cloud/stream/app/gemfire/config/GemfireSslProperties.java:
```diff
@@ -0,0 +1,93 @@
+ * @author Christian Tzolov
+ */
+@ConfigurationProperties("gemfire.security.ssl")
+public class GemfireSslProperties {
```
Don't we need to have some validation on the properties to be sure that end-user has provided all the required?
For example when we have a `truststoreUri` then we need to have a `sslKeystorePassword` and so on.

tzolov replied:
I guess we can add few cross-field invariants: 
- Those invariants should be enforced only when  the `truststoreUri` is not null.
- The sslKeystorePassword and sslTruststorePassword can be either empty or both set.
- The userHomeDirectory and storeFileName should not be empty!

artembilan commented on spring-cloud-starter-stream-common-gemfire/src/main/java/org/springframework/cloud/stream/app/gemfire/config/GemfireClientCacheConfiguration.java:
```diff
@@ -15,42 +15,91 @@
+		String trustKeyStorePath = this.resolveTrustedKeystore(sslProperties);
+		PropertiesBuilder pb = new PropertiesBuilder();
+		pb.setProperty("ssl-enabled-components", "server,locator");
+		pb.setProperty("ssl-keystore", trustKeyStorePath);
```
Don't you miss here a separate `keyStorePath` property?
This is not recommended to keep a trusted certificates and secret keys in the same store.

More over you have a separate properties for passwords, what doesn't sound logical if file is the same...

tzolov replied:
The initial impl. was based on the [Geode sample](https://gemfire.docs.pivotal.io/95/geode/managing/security/ssl_example.html) but i will expand it to better support the available ssl properties: https://gemfire.docs.pivotal.io/95/geode/managing/security/implementing_ssl.html
Furthermore, if needed it will be still possible to reuse the same file as both trust/key store if the same URI is used for both.

artembilan commented on spring-cloud-starter-stream-common-gemfire/src/main/java/org/springframework/cloud/stream/app/gemfire/config/GemfireSslProperties.java:
```diff
@@ -0,0 +1,93 @@
+	/**
+	 * Name of the trust store file copied in the local file system.
+	 */
+	private String storeFileName = "trusted.keystore";
```
Does really this local file name matter?
We definitely need a local file because SSL doesn't work other way, but I don't see too much value to expose this as a configuration property.

I can be wrong if you are able to convince me 😉

tzolov replied:
Correct, i will convert it into a static constant.

artembilan commented on spring-cloud-starter-stream-sink-gemfire/src/test/java/org/springframework/cloud/stream/app/gemfire/sink/bar/GemfireSinkConfigurationTests.java:
```diff
@@ -13,7 +13,7 @@
  */
 
-package org.springframework.cloud.stream.app.gemfire.sink;
+package org.springframework.cloud.stream.app.gemfire.sink.bar;
```
Huh? What does this `bar` package mean?

tzolov replied:
my bad. Missed to clean this after my experiments to make both GemfireSinkIntegrationTests and SslGemfireSinkIntegrationTests work together.

artembilan commented on spring-cloud-starter-stream-sink-gemfire/src/test/java/org/springframework/cloud/stream/app/gemfire/sink/bar/GemfireSinkConfigurationTests.java:
```diff
@@ -38,9 +40,11 @@
+		"gemfire.pool.hostAddresses=localhost:42424", "gemfire.pool.connectType=server",
+		"spring.cloud.stream.default.binder=test" },
+		classes = { GemfireSinkConfiguration.class })
+@EnableConfigurationProperties({ GemfireSinkProperties.class, GemfireSecurityProperties.class })
```
I believe we don't need this here since we rely on the `GemfireSinkConfiguration` in the `@SpringBootTest`

tzolov replied:
will fix this

artembilan commented on spring-cloud-starter-stream-sink-gemfire/src/test/java/org/springframework/cloud/stream/app/gemfire/sink/normal/GemfireSinkIntegrationTests.java:
```diff
@@ -58,8 +50,13 @@
+				"spring.cloud.stream.default.binder=test"
+		},
+        classes = {GemfireSinkConfiguration.class})
 @DirtiesContext(methodMode = DirtiesContext.MethodMode.BEFORE_METHOD)
```
This is wrong. I removed the `methodMode` from all the `@DirtiesContext`.

tzolov replied:
this as well

artembilan commented on spring-cloud-starter-stream-sink-gemfire/src/test/java/org/springframework/cloud/stream/app/gemfire/sink/normal/GemfireSinkIntegrationTests.java:
```diff
@@ -126,6 +98,7 @@ public void testByteArrayJsonPayload() {
 			assertThat(transformer.toString(region.get("key")), equalTo(messageBody));
 		}
 
+		@Ignore
```
After un`@Ignore` these test and running together with the `GemfireSinkIntegrationTests` I get this exception:
```
Caused by: org.apache.geode.cache.client.ServerOperationException: remote server on 10.0.0.134(11316:loner):62744:7a355dcf: org.apache.geode.security.AuthenticationRequiredException: Server expecting SSL connection
	at org.apache.geode.cache.client.internal.pooling.ConnectionManagerImpl.borrowConnection(ConnectionManagerImpl.java:246)
	at org.apache.geode.cache.client.internal.OpExecutorImpl.execute(OpExecutorImpl.java:148)
	at org.apache.geode.cache.client.internal.OpExecutorImpl.execute(OpExecutorImpl.java:127)
	at org.apache.geode.cache.client.internal.PoolImpl.execute(PoolImpl.java:782)
	at org.apache.geode.cache.client.internal.PutAllOp.execute(PutAllOp.java:100)
	at org.apache.geode.cache.client.internal.ServerRegionProxy.putAll(ServerRegionProxy.java:604)
	at org.apache.geode.internal.cache.LocalRegion.basicPutAll(LocalRegion.java:9515)
	at org.apache.geode.internal.cache.LocalRegion.putAll(LocalRegion.java:9448)
	at org.apache.geode.internal.cache.LocalRegion.putAll(LocalRegion.java:9460)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.data.gemfire.GemfireTemplate$RegionCloseSuppressingInvocationHandler.invoke(GemfireTemplate.java:534)
	at com.sun.proxy.$Proxy94.putAll(Unknown Source)
	at org.springframework.integration.gemfire.outbound.CacheWritingMessageHandler$1.doInGemfire(CacheWritingMessageHandler.java:96)
	at org.springframework.data.gemfire.GemfireTemplate.execute(GemfireTemplate.java:459)
	... 65 more
```

I believe you get something similar, but still it fully unclear to me what is wrong. Something like we have a server in SSL mode, but our client connects without 🤷‍♂️

tzolov replied:
You can run both tests one by one, but apparently not both together. 
They obviously interfere and i was not able to track the cause of it. Furthermore the existing geode integration testing infrastructure looks bit outdated and unnecessarily complicated. I believe we can improve this so i created https://github.com/spring-cloud-stream-app-starters/gemfire/issues/19.

tzolov committed:
Add separate keystore and truststore URI. Add property validation

tzolov committed:
Resolve issues with test annotations

tzolov commented:
@artembilan , i've pushed couple of commits to address most of the issues above. Still the interfering ITests is not solved.

artembilan commented on spring-cloud-starter-stream-common-gemfire/src/main/java/org/springframework/cloud/stream/app/gemfire/config/GemfireClientCacheConfiguration.java:
```diff
@@ -22,11 +22,15 @@
 import org.springframework.util.StringUtils;
 
+import static org.springframework.cloud.stream.app.gemfire.config.GemfireSslProperties.LOCAL_KEYSTORE_FILE_NAME;
+import static org.springframework.cloud.stream.app.gemfire.config.GemfireSslProperties.LOCAL_TRUSTSTORE_FILE_NAME;
```
Our policy do not have `static` imports in the `main` code.

I think it's not critical, but still: we develop here Spring.

artembilan commented on spring-cloud-starter-stream-common-gemfire/src/main/java/org/springframework/cloud/stream/app/gemfire/config/GemfireSslProperties.java:
```diff
@@ -15,41 +15,74 @@
+	/**
+	 * Local directory to store the truststore and keystore files copied form the remote truststoreUri and getKeystoreUri uris.
+	 */
+	@NotBlank
```
Not sure if this works, but in other places we use validation annotations on getters.

artembilan replied:
Works well. Resolving...

artembilan commented on spring-cloud-starter-stream-sink-gemfire/src/test/java/org/springframework/cloud/stream/app/gemfire/sink/normal/GemfireSinkIntegrationTests.java:
```diff
@@ -56,8 +56,8 @@
-        classes = {GemfireSinkConfiguration.class})
-@DirtiesContext(methodMode = DirtiesContext.MethodMode.BEFORE_METHOD)
+		classes = { GemfireSinkConfiguration.class })
+@DirtiesContext(classMode = DirtiesContext.ClassMode.AFTER_EACH_TEST_METHOD)
```
Actually it works even just with the default after class.
I mean that we don't need this extra control after each method.

artembilan commented on spring-cloud-starter-stream-sink-gemfire/src/test/java/org/springframework/cloud/stream/app/gemfire/sink/ssl/SslGemfireSinkIntegrationTests.java:
```diff
@@ -60,7 +60,7 @@
 		},
 		classes = {GemfireSinkConfiguration.class})
-@DirtiesContext(methodMode = DirtiesContext.MethodMode.BEFORE_METHOD)
+@DirtiesContext(classMode = DirtiesContext.ClassMode.AFTER_EACH_TEST_METHOD)
```
DITTO.

artembilan reviewed (changes requested):
I'm fine to stay with the ignored test for time being since it doesn't look like a problem from our side. Although we have all those support classes for the separate Gemfilre JVM...

MAy we consider to ignore regular tests, but not those with SSL?

Thanks

tzolov committed:
Mark constant as final

tzolov committed:
Use default @DirtiesContext attributes. Adhere to the static imports policy

artembilan commented:
@tzolov ,

Let me know if you are going to do anything else here, otherwise it is good to go for merging.

tzolov committed:
- Remove redundant code.
 - Improve ITest ignore msg description.
 - Add additional non-blank SSL properties tests.

tzolov committed:
Add doc for SSL activation

tzolov committed:
Ajdust the ingored statements

tzolov commented:
@artembilan have somehow wrongly assumed that this issues resolved.  
Just pushed the latest changes. Since it is been a while you may consider re-reviewing it again :(

artembilan commented on spring-cloud-starter-stream-sink-gemfire/README.adoc:
```diff
@@ -3,6 +3,10 @@
 
+To enable SSL communication between Geode Sink and the Geode cluster you need to provide the URIs of the
+Keystore and Truststore files using the `gemfire.security.ssl.keystore-uri` and `gemfire.security.ssl.truststore-uri` properties.
+(If a single file is ued for both stores then point both URIs to it).
```
typo: is u**s**ed

artembilan commented on spring-cloud-starter-stream-common-gemfire/src/main/java/org/springframework/cloud/stream/app/gemfire/config/GemfireSslProperties.java:
```diff
@@ -0,0 +1,167 @@
+	public static final String LOCAL_KEYSTORE_FILE_NAME = "keystore.keystore";
+
+	/**
+	 * Local directory to store the truststore and keystore files copied form the remote truststoreUri and getKeystoreUri uris.
```
I believe this is typo: truststore and keystore uris

Correct me if I'm wrong.

tzolov replied:
I've re-wrote the description. Hopefully it is cleaner now

artembilan commented on spring-cloud-starter-stream-common-gemfire/src/main/java/org/springframework/cloud/stream/app/gemfire/config/GemfireSslProperties.java:
```diff
@@ -0,0 +1,167 @@
+	private String userHomeDirectory = USER_HOME_DIRECTORY;
+
+	/**
+	 * Location of the pre-created truststore URI to be used for connecting to the Geode cluster
```
Pay attention how all the configuration properties JavaDocs should end with the period and should not use JavaDoc tag, like you do in the `truststoreType` description.

That's Spring Boot requirements.

artembilan reviewed (changes requested):
That's all. Just some docs polishing necessary yet.

tzolov committed:
Doc related changeds to address the review requests
